File: backend/app/models/production.py
from __future__ import annotations

from datetime import date, datetime
from enum import Enum

from sqlalchemy import ForeignKey, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.app.core.db_base import Base


class ProductionOrderStatus(str, Enum):
    pending = "pending"
    awaiting_resources = "awaiting_resources"
    ready_to_work = "ready_to_work"
    in_progress = "in_progress"
    completed = "completed"
    partially_transferred = "partially_transferred"
    transferred = "transferred"


class ProductionSupplyType(str, Enum):
    raw_material = "raw_material"
    consumable = "consumable"
    finished_goods_receipt = "finished_goods_receipt"


class ProductionSupplyStatus(str, Enum):
    requested = "requested"
    in_progress = "in_progress"
    completed = "completed"


class ProductionChzStatus(str, Enum):
    requested = "requested"
    acknowledged = "acknowledged"
    ready = "ready"
    cancelled = "cancelled"


class ProductionTaskType(str, Enum):
    packaging = "packaging"
    unpacking = "unpacking"
    trim_cuffs = "trim_cuffs"
    warehouse_help = "warehouse_help"
    defect_sorting = "defect_sorting"
    repacking = "repacking"
    cleaning = "cleaning"


class ProductionOrder(Base):
    __tablename__ = "production_orders"

    id: Mapped[int] = mapped_column(primary_key=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    task_type: Mapped[ProductionTaskType] = mapped_column(
        String(32),
        default=ProductionTaskType.packaging,
        nullable=False,
    )
    status: Mapped[ProductionOrderStatus] = mapped_column(
        String(32),
        default=ProductionOrderStatus.pending,
        nullable=False,
    )
    priority: Mapped[int] = mapped_column(default=5, nullable=False)
    comment: Mapped[str | None] = mapped_column(String(2000), nullable=True)
    related_order_id: Mapped[int | None] = mapped_column(ForeignKey("orders.id", ondelete="SET NULL"), nullable=True)
    batch_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    production_date: Mapped[date | None] = mapped_column(nullable=True)
    created_by_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    brigadier_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now(), onupdate=func.now())

    related_order: Mapped["Order | None"] = relationship("Order")
    created_by: Mapped["User | None"] = relationship("User", foreign_keys=[created_by_user_id])
    brigadier: Mapped["User | None"] = relationship("User", foreign_keys=[brigadier_user_id])
    items: Mapped[list["ProductionOrderItem"]] = relationship(
        "ProductionOrderItem",
        back_populates="production_order",
        cascade="all, delete-orphan",
    )
    supply_requests: Mapped[list["ProductionSupplyRequest"]] = relationship(
        "ProductionSupplyRequest",
        back_populates="production_order",
        cascade="all, delete-orphan",
    )
    chz_requests: Mapped[list["ProductionChzRequest"]] = relationship(
        "ProductionChzRequest",
        back_populates="production_order",
        cascade="all, delete-orphan",
    )
    labor_entries: Mapped[list["ProductionLaborEntry"]] = relationship(
        "ProductionLaborEntry",
        back_populates="production_order",
        cascade="all, delete-orphan",
    )


class ProductionOrderItem(Base):
    __tablename__ = "production_order_items"

    id: Mapped[int] = mapped_column(primary_key=True)
    production_order_id: Mapped[int] = mapped_column(ForeignKey("production_orders.id", ondelete="CASCADE"))
    item_id: Mapped[int] = mapped_column(ForeignKey("items.id", ondelete="RESTRICT"))
    pairs_quantity: Mapped[int] = mapped_column(nullable=False)
    produced_pairs: Mapped[int] = mapped_column(default=0, nullable=False)
    transferred_pairs: Mapped[int] = mapped_column(default=0, nullable=False)
    
    item_size: Mapped[str | None] = mapped_column(String(50), nullable=True)
    item_color: Mapped[str | None] = mapped_column(String(50), nullable=True)
    
    # ✅ НОВЫЕ ПОЛЯ для партии и даты на уровне позиции
    batch_number: Mapped[str | None] = mapped_column(String(50), nullable=True)
    production_date: Mapped[date | None] = mapped_column(nullable=True)

    production_order: Mapped["ProductionOrder"] = relationship("ProductionOrder", back_populates="items")
    item: Mapped["Item"] = relationship("Item")
    transfers: Mapped[list["ProductionTransfer"]] = relationship(
        "ProductionTransfer",
        back_populates="production_order_item",
        cascade="all, delete-orphan",
    )
    history: Mapped[list["ProductionHistory"]] = relationship(
        "ProductionHistory",
        back_populates="production_order_item",
        cascade="all, delete-orphan",
        order_by="ProductionHistory.created_at.desc()"
    )

class ProductionSupplyRequest(Base):
    __tablename__ = "production_supply_requests"

    id: Mapped[int] = mapped_column(primary_key=True)
    production_order_id: Mapped[int] = mapped_column(ForeignKey("production_orders.id", ondelete="CASCADE"))
    request_type: Mapped[ProductionSupplyType] = mapped_column(String(32), nullable=False)
    status: Mapped[ProductionSupplyStatus] = mapped_column(
        String(32),
        default=ProductionSupplyStatus.requested,
        nullable=False,
    )
    comment: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(server_default=func.now(), onupdate=func.now())

    production_order: Mapped["ProductionOrder"] = relationship("ProductionOrder", back_populates="supply_requests")
    items: Mapped[list["ProductionSupplyRequestItem"]] = relationship(
        "ProductionSupplyRequestItem",
        back_populates="request",
        cascade="all, delete-orphan",
    )


class ProductionSupplyRequestItem(Base):
    __tablename__ = "production_supply_request_items"

    id: Mapped[int] = mapped_column(primary_key=True)
    request_id: Mapped[int] = mapped_column(ForeignKey("production_supply_requests.id", ondelete="CASCADE"))
    item_id: Mapped[int] = mapped_column(ForeignKey("items.id", ondelete="RESTRICT"))
    production_order_item_id: Mapped[int | None] = mapped_column(
        ForeignKey("production_order_items.id", ondelete="SET NULL"),
        nullable=True,
    )
    quantity: Mapped[int] = mapped_column(nullable=False)
    fulfilled_quantity: Mapped[int] = mapped_column(default=0, nullable=False)
    consumed_quantity: Mapped[int] = mapped_column(default=0, nullable=False)
    size: Mapped[str | None] = mapped_column(String(50), nullable=True)
    manufacturer: Mapped[str | None] = mapped_column(String(255), nullable=True)
    selected_stock_id: Mapped[int | None] = mapped_column(ForeignKey("stocks.id", ondelete="SET NULL"), nullable=True)
    selected_cell_id: Mapped[int | None] = mapped_column(ForeignKey("cells.id", ondelete="SET NULL"), nullable=True)

    request: Mapped["ProductionSupplyRequest"] = relationship("ProductionSupplyRequest", back_populates="items")
    item: Mapped["Item"] = relationship("Item")
    production_order_item: Mapped["ProductionOrderItem | None"] = relationship("ProductionOrderItem")
    selected_stock: Mapped["Stock | None"] = relationship("Stock")
    selected_cell: Mapped["Cell | None"] = relationship("Cell")


class ProductionChzRequest(Base):
    __tablename__ = "production_chz_requests"

    id: Mapped[int] = mapped_column(primary_key=True)
    production_order_id: Mapped[int] = mapped_column(ForeignKey("production_orders.id", ondelete="CASCADE"))
    requested_by_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    status: Mapped[ProductionChzStatus] = mapped_column(
        String(32),
        default=ProductionChzStatus.requested,
        nullable=False,
    )
    is_active: Mapped[bool] = mapped_column(default=True, nullable=False)
    comment: Mapped[str | None] = mapped_column(String(1000), nullable=True)
    external_request_id: Mapped[str | None] = mapped_column(String(255), nullable=True)
    requested_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)
    acknowledged_at: Mapped[datetime | None] = mapped_column(nullable=True)
    ready_at: Mapped[datetime | None] = mapped_column(nullable=True)

    production_order: Mapped["ProductionOrder"] = relationship("ProductionOrder", back_populates="chz_requests")
    requested_by: Mapped["User | None"] = relationship("User")
    items: Mapped[list["ProductionChzRequestItem"]] = relationship(
        "ProductionChzRequestItem",
        back_populates="request",
        cascade="all, delete-orphan",
    )

    @property
    def order_name(self) -> str | None:
        return self.production_order.name if self.production_order else None

    @property
    def requested_by_username(self) -> str | None:
        return self.requested_by.username if self.requested_by else None

    @property
    def request_type(self) -> str:
        return "production"


class ProductionChzRequestItem(Base):
    __tablename__ = "production_chz_request_items"

    id: Mapped[int] = mapped_column(primary_key=True)
    request_id: Mapped[int] = mapped_column(ForeignKey("production_chz_requests.id", ondelete="CASCADE"))
    production_order_item_id: Mapped[int | None] = mapped_column(ForeignKey("production_order_items.id", ondelete="SET NULL"), nullable=True)
    item_id: Mapped[int | None] = mapped_column(ForeignKey("items.id", ondelete="SET NULL"), nullable=True)
    pairs_quantity: Mapped[int] = mapped_column(nullable=False)
    item_title: Mapped[str] = mapped_column(String(255), nullable=False)
    item_size: Mapped[str | None] = mapped_column(String(50), nullable=True)
    item_color: Mapped[str | None] = mapped_column(String(50), nullable=True)
    batch_number: Mapped[str | None] = mapped_column(String(50), nullable=True)

    request: Mapped["ProductionChzRequest"] = relationship("ProductionChzRequest", back_populates="items")
    item: Mapped["Item | None"] = relationship("Item")
    production_order_item: Mapped["ProductionOrderItem | None"] = relationship("ProductionOrderItem")


class ProductionTransfer(Base):
    __tablename__ = "production_transfers"

    id: Mapped[int] = mapped_column(primary_key=True)
    production_order_item_id: Mapped[int] = mapped_column(ForeignKey("production_order_items.id", ondelete="CASCADE"))
    stock_id: Mapped[int | None] = mapped_column(ForeignKey("stocks.id", ondelete="SET NULL"), nullable=True)
    cell_id: Mapped[int | None] = mapped_column(ForeignKey("cells.id", ondelete="SET NULL"), nullable=True)
    pairs_quantity: Mapped[int] = mapped_column(nullable=False)
    created_by_user_id: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"), nullable=True)
    transferred_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)

    production_order_item: Mapped["ProductionOrderItem"] = relationship("ProductionOrderItem", back_populates="transfers")
    stock: Mapped["Stock | None"] = relationship("Stock")
    cell: Mapped["Cell | None"] = relationship("Cell")
    created_by: Mapped["User | None"] = relationship("User")


class ProductionHistory(Base):
    __tablename__ = "production_history"

    id: Mapped[int] = mapped_column(primary_key=True)
    production_order_item_id: Mapped[int] = mapped_column(
        ForeignKey("production_order_items.id", ondelete="CASCADE"),
        nullable=False
    )
    old_produced_pairs: Mapped[int] = mapped_column(nullable=False)
    new_produced_pairs: Mapped[int] = mapped_column(nullable=False)
    changed_by_user_id: Mapped[int | None] = mapped_column(
        ForeignKey("users.id", ondelete="SET NULL"),
        nullable=True
    )
    comment: Mapped[str | None] = mapped_column(String(500), nullable=True)
    created_at: Mapped[datetime] = mapped_column(server_default=func.now(), nullable=False)

    production_order_item: Mapped["ProductionOrderItem"] = relationship(
        "ProductionOrderItem",
        back_populates="history"
    )
    changed_by: Mapped["User | None"] = relationship("User")

File: frontend/src/lib/production.ts
// lib/production.ts

import type {
  EmployeeDepartment,
  ProductionChzStatus,
  ProductionOrder,
  ProductionOrderItem,
  ProductionSupplyStatus,
  ProductionSupplyType,
  ProductionTaskType,
} from '@/types/wms'

export type ProductionTone = 'success' | 'warning' | 'danger' | 'neutral' | 'secondary' | 'info'

export type ProductionWarehouseTask = {
  key: string
  orderId: number
  orderName: string
  priority: number
  createdAt: string
  requestId: number
  requestItemId: number
  requestType: ProductionSupplyType
  requestStatus: ProductionSupplyStatus
  requestComment?: string | null
  productionOrderItemId?: number | null
  itemId: number
  itemTitle: string
  itemSize?: string | null
  itemColor?: string | null      // ✅ ДОБАВЛЯЕМ цвет
  batchNumber?: string | null
  remainingQuantity: number
}

export function productionStatusLabel(status: string) {
  switch (status) {
    case 'pending':
      return 'Ожидает'
    case 'awaiting_resources':
      return 'Ожидает ресурсы'
    case 'ready_to_work':
      return 'Готов к работе'
    case 'in_progress':
      return 'В работе'
    case 'completed':
      return 'Выполнено'
    case 'partially_transferred':
      return 'Частично передано'
    case 'transferred':
      return 'Передано на склад'
    default:
      return status
  }
}

export function productionStatusTone(status: string): ProductionTone {
  switch (status) {
    case 'ready_to_work':
      return 'info'
    case 'in_progress':
      return 'warning'
    case 'completed':
    case 'transferred':
      return 'success'
    case 'partially_transferred':
      return 'secondary'
    default:
      return 'neutral'
  }
}

export function productionSupplyTypeLabel(type: ProductionSupplyType) {
  switch (type) {
    case 'raw_material':
      return 'Сырье'
    case 'consumable':
      return 'Расходники'
    case 'finished_goods_receipt':
      return 'Приемка ГП'
    default:
      return type
  }
}

export function productionSupplyTypeUnitLabel(type: ProductionSupplyType) {
  if (type === 'consumable') return 'шт'
  return 'пар'
}

export function productionSupplyStatusLabel(status: ProductionSupplyStatus) {
  switch (status) {
    case 'requested':
      return 'Ожидает'
    case 'in_progress':
      return 'В работе'
    case 'completed':
      return 'Выполнено'
    default:
      return status
  }
}

export function productionSupplyStatusTone(status: ProductionSupplyStatus): ProductionTone {
  switch (status) {
    case 'requested':
      return 'warning'
    case 'in_progress':
      return 'info'
    case 'completed':
      return 'success'
    default:
      return 'neutral'
  }
}

export function productionChzStatusLabel(status: ProductionChzStatus | string) {
  switch (status) {
    case 'requested':
      return 'Запрошен'
    case 'acknowledged':
      return 'Принят оператором'
    case 'ready':
      return 'Коды готовы'
    case 'cancelled':
      return 'Отменен'
    default:
      return status
  }
}

export function productionChzStatusTone(status: ProductionChzStatus | string): ProductionTone {
  switch (status) {
    case 'ready':
      return 'success'
    case 'acknowledged':
      return 'info'
    case 'requested':
      return 'warning'
    case 'cancelled':
      return 'danger'
    default:
      return 'neutral'
  }
}

export function formatProductionItemLabel(item: Pick<ProductionOrderItem, 'item_title' | 'item_size'>) {
  return item.item_size ? `${item.item_title} · р. ${item.item_size}` : item.item_title
}

export function getProductionItemTransferablePairs(item: Pick<ProductionOrderItem, 'pairs_quantity' | 'produced_pairs'>) {
  return item.produced_pairs > 0 ? item.produced_pairs : item.pairs_quantity
}

export function sumProductionPairs(
  order: ProductionOrder,
  key: 'pairs_quantity' | 'produced_pairs' | 'transferred_pairs',
) {
  return order.items.reduce((total, item) => total + item[key], 0)
}

export function getProductionProgress(order: ProductionOrder) {
  const total = sumProductionPairs(order, 'pairs_quantity')
  const produced = sumProductionPairs(order, 'produced_pairs')
  return {
    produced,
    total,
    percent: total > 0 ? Math.round((produced / total) * 100) : 0,
  }
}

export function buildProductionTaskKey(requestId: number, requestItemId: number) {
  return `request-${requestId}-item-${requestItemId}`
}

export function getProductionWarehouseTasks(orders: ProductionOrder[]): ProductionWarehouseTask[] {
  return orders
    .flatMap((order) =>
      order.supply_requests.flatMap((request) =>
        request.items.flatMap((item) => {
          const remainingQuantity = Math.max(item.quantity - item.fulfilled_quantity, 0)
          if (remainingQuantity <= 0 || request.status === 'completed') return []

          // Находим связанную позицию заказа
          const relatedOrderItem = order.items.find((candidate) => candidate.id === item.production_order_item_id)

          // Определяем размер, цвет и партию в зависимости от типа задания
          let itemSize = item.size ?? item.item_size ?? relatedOrderItem?.item_size ?? null
          let itemColor = relatedOrderItem?.item_color ?? null
          let batchNumber = null

          if (request.request_type === 'finished_goods_receipt' && relatedOrderItem) {
            // ✅ Для приёмки ГП берём данные из позиции заказа
            itemSize = relatedOrderItem.item_size ?? itemSize
            itemColor = relatedOrderItem.item_color ?? null
            batchNumber = relatedOrderItem.batch_number ?? order.batch_number ?? null
          } else {
            // Для сырья и расходников используем партию заказа
            batchNumber = order.batch_number ?? null
          }

          return [
            {
              key: buildProductionTaskKey(request.id, item.id),
              orderId: order.id,
              orderName: order.name,
              priority: order.priority,
              createdAt: request.created_at,
              requestId: request.id,
              requestItemId: item.id,
              requestType: request.request_type,
              requestStatus: request.status,
              requestComment: request.comment,
              productionOrderItemId: item.production_order_item_id,
              itemId: item.item_id,
              itemTitle: item.item_title,
              itemSize: itemSize,
              itemColor: itemColor,           // ✅ цвет теперь передаётся
              batchNumber: batchNumber,
              remainingQuantity,
            } satisfies ProductionWarehouseTask,
          ]
        }),
      ),
    )
    .sort((left, right) => {
      if (left.priority !== right.priority) return right.priority - left.priority
      return left.createdAt < right.createdAt ? 1 : -1
    })
}

export function productionTaskTypeLabel(type: ProductionTaskType | string) {
  switch (type) {
    case 'packaging':
      return 'Упаковка'
    case 'unpacking':
      return 'Распаковка'
    case 'trim_cuffs':
      return 'Обрезка венчиков'
    case 'warehouse_help':
      return 'Помощь на складе'
    case 'defect_sorting':
      return 'Переборка брака'
    case 'repacking':
      return 'Переупаковка'
    case 'cleaning':
      return 'Уборка'
    default:
      return type
  }
}

export function employeeDepartmentLabel(department?: EmployeeDepartment | string | null) {
  switch (department) {
    case 'production':
      return 'Производство'
    case 'warehouse':
      return 'Склад'
    case 'other':
      return 'Другое'
    default:
      return department || '—'
  }
}

export function productionBrigadierChzStatusLabel(status?: ProductionChzStatus | string | null) {
  switch (status) {
    case 'requested':
      return 'ЧЗ запрошен'
    case 'acknowledged':
      return 'В работе'
    case 'ready':
      return 'Готов'
    case 'cancelled':
      return 'Отменен'
    default:
      return status || 'Нет запроса'
  }
}
